nodejs/device/src/index.ts

```diff
@@ -0,0 +1,318 @@
+  }
+
+  private registerMessageListener() {
+    this.client.on('message', msg => {
+      let obj;
+      try {
+        obj = JSON.parse(msg.getData());
+      }
+      catch {
+        this.rejectC2d(msg);
+        return console.warn('Failed to parse received C2D message. Rejecting.');
+      }
+
+      if (!obj.thief || obj.thief.runId !== runId || obj.thief.serviceInstanceId !== this.serviceInstanceId) {
+        this.rejectC2d(msg);
+        console.warn("C2D received, but it's not for us: %j. Rejecting.", obj);
+      }
+      else if (obj.thief.cmd === 'serviceAckResponse') {
+        this.handleServiceAckResponseMessage(obj, msg);
+      }
+      else if (obj.thief.cmd === 'testC2d') {
+        this.handleTestC2dMessage(obj, msg);
+      }
+      else {
+        this.rejectC2d(msg);
+        console.warn('Unknown command received: %j. Rejecting', obj);
+      }
+    });
+  }
+
+  private startC2dMessageSending(): Promise<void> {
+    const patch = {
+      thief: {
+        testControl: {
+          c2d: {
+            messageIntervalInSeconds: this.settings.receiveC2dIntervalInSeconds,
+            send: true
+          }
+        }
+      } 
+    }
+    console.info('Enabling C2D message testing: %j', patch);
+    return promisify(this.twin.properties.reported.update)(patch);
+  }
+
+  private startTelemetrySending() {
+    this.telemetrySendingInterval = setInterval(() => {
+      if (this.serviceAckWaitList.size > this.settings.sendMessageAllowedFailureCount) {
+        this.stopDeviceApp(`The number of service acks being waited on exceeds sendMessageAllowedFailureCount of ${this.settings.sendMessageAllowedFailureCount}.`);
+      }
+
+      const serviceAckId = uuidv4();
+      const message = new Message(JSON.stringify({
+        thief: {
+          cmd: 'serviceAckRequest',
+          serviceInstanceId: this.serviceInstanceId,
+          runId: runId,
+          serviceAckId: serviceAckId,
+          serviceAckType: 'telemetry'
+        }
+      }));
+      this.client.sendEvent(message)
+        .catch(() => {console.warn('Error sending message with serviceAckId: ' + serviceAckId)})
+        .finally(() => {this.serviceAckWaitList.add(serviceAckId)});
+    }, 1000 * this.settings.sendMessageOperationsPerSecond);
+  }
+
+  private startTestOperations() {
+    this.registerMessageListener();
+    this.startC2dMessageSending().catch(e => {throw new Error('Starting C2D message sending failed: ' + e.message)});
+    this.startTelemetrySending();
+  }
+
+  stopDeviceApp(reason?: string | Error) {
+    console.warn('Stopping device app' + (reason ? `: ${reason}`: ''));
+    clearInterval(this.telemetrySendingInterval);
+    clearTimeout(this.pairingAttemptTimer);
+    clearTimeout(this.pairingTimer);
+    if (this.twin) {
+      this.twin.removeAllListeners();
+    }
+    if (this.client) {
+      this.client.removeAllListeners();
+      this.client.close().catch(e => console.error(`Error closing client: ${e}`));
+    }
```

Comment:
If I'm reading this right, this calls close as an async operation and returns.  If close fails, it outputs to the console.  There is no way, from outside this function, to know _when_ the close operation succeeds.
Since it's calling an async function( `client.close()`), should `stopDeviceApp` also be an async function?

Reply:
I don't know if this is canon on the Node team, but I think every function, except for simple ones (getters/setters/transformers) should either accept a callback or return a promise.  (I assume that putting the `async` keyword in front of a function causes it to return a promise, right?)

---
In reply to: [577844534](https://github.com/Azure/iot-sdk-longhaul/pull/61#discussion_r577844534) [](ancestors = 577844534)

Reply:
Well the goal of `stopDeviceApp()` is that it clears everything from the event loop, causing Node to exit.
When it succeeds, you would know because Node exited without printing an error. Though it should print a success message anyway. I will do that. 

I think this also explains why returning a promise from the function isn't that useful.